stan/math/gpu/kernel_cl.hpp

```diff
@@ -29,24 +29,46 @@ map_base_opts base_opts
        {"UPPER", static_cast<int>(TriangularViewGPU::Upper)},
        {"ENTIRE", static_cast<int>(TriangularViewGPU::Entire)},
        {"UPPER_TO_LOWER", static_cast<int>(TriangularMapGPU::UpperToLower)},
-       {"LOWER_TO_UPPER", static_cast<int>(TriangularMapGPU::LowerToUpper)}};
+       {"LOWER_TO_UPPER", static_cast<int>(TriangularMapGPU::LowerToUpper)},
+       {"WORK_PER_WI_MULT", 8},
+       {"WG_SIZE_MULT", 32},
+       {"WG_SIZE_MULT_SELF_TRANS", 32},
+       {"WORK_PER_WI_MULT_SELF_TRANS", 4}};
 
 auto compile_kernel(const char* name, const char* source) {
+  size_t max_wg_size = opencl_context.max_workgroup_size();
+  int wg_size_sqrt = static_cast<int>(sqrt(static_cast<double>(max_wg_size)));
+  // Does a compile time check of the maximum allowed
+  // dimension of a square workgroup size
+  // WG size of (32,32) works on all recent GPU but would fail on some
+  // older integrated GPUs or CPUs
+  if (wg_size_sqrt < base_opts["WG_SIZE_MULT"]){
```

Comment:
Do we want to do this here? I dont have any better ideas right now.

Reply:
Hmmm. Can we make this stuff live on the opencl_context and be computed just once when the context is initialized? It can then expose the options map as part of it's API perhaps?

Reply:
Tnx! Seems like a good idea.